crates/sui-source-validation/src/toolchain.rs

```diff
@@ -1,440 +1,540 @@
-//         flavor: Flavor::Sui,      /* does not matter, unused for current_toolchain */
-//     }
-// }
-//
-// pub(crate) fn legacy_toolchain() -> ToolchainVersion {
-//     ToolchainVersion {
-//         compiler_version: LEGACY_COMPILER_VERSION.into(),
-//         edition: Edition::LEGACY,
-//         flavor: Flavor::Sui,
-//     }
-// }
-//
+
+pub(crate) const CURRENT_COMPILER_VERSION: &str = env!("CARGO_PKG_VERSION");
+const LEGACY_COMPILER_VERSION: &str = CURRENT_COMPILER_VERSION; // TODO: update this when Move 2024 is released
+const PRE_TOOLCHAIN_MOVE_LOCK_VERSION: u16 = 0; // Used to detect lockfiles pre-toolchain versioning support
+const CANONICAL_UNIX_BINARY_NAME: &str = "sui";
+const CANONICAL_WIN_BINARY_NAME: &str = "sui.exe";
+
+/// Lock file version written by this version of the compiler.  Backwards compatibility is
+/// guaranteed (the compiler can read lock files with older versions), forward compatibility is not
+/// (the compiler will fail to read lock files at newer versions).
+///
+/// V0: Base version.
+/// V1: Adds toolchain versioning support.
+/// V2: Adds support for managing addresses on package publish and upgrades.
+/// V3: Renames dependency `name` field to `id` and adds a `name` field to store the name from the manifest.
+pub const VERSION: u16 = 3;
+
+#[derive(Serialize, Deserialize)]
+pub struct LockfileHeader {
+    pub version: u16,
+}
+
+// TODO: pkg-alt this needs to work with both old style and new style formats. Particularly, for
+// the new pkg system, the toolchain version is in the Published.toml file, or Pub.env.toml file.
+#[derive(Serialize, Deserialize, Debug)]
+pub struct ToolchainVersion {
+    /// The Move compiler version used to compile this package.
+    #[serde(rename = "compiler-version")]
+    pub compiler_version: String,
+    /// The Move compiler configuration used to compile this package.
+    pub edition: Edition,
+    pub flavor: Flavor,
+}
+
+#[derive(Serialize, Deserialize)]
+struct Schema<T> {
+    #[serde(rename = "move")]
+    move_: T,
+}
+
+impl LockfileHeader {
+    /// Read lock file header after verifying that the version of the lock is not newer than the version
+    /// supported by this library.
+    pub fn read(lock: &mut impl Read) -> Result<Self> {
+        let contents = {
+            let mut buf = String::new();
+            lock.read_to_string(&mut buf).context("Reading lock file")?;
+            buf
+        };
+        Self::from_str(&contents)
+    }
+
+    fn from_str(contents: &str) -> Result<Self> {
+        let Schema { move_: header } =
+            toml::de::from_str::<Schema<Self>>(contents).context("Deserializing lock header")?;
+
+        if header.version != VERSION {
+            bail!(
+                "Lock file format mismatch, expected version {}, found {}",
+                VERSION,
+                header.version
+            );
+        }
+
+        Ok(header)
+    }
+}
+
+impl ToolchainVersion {
```

Comment:
It seems like a lot of this code is duplicated from move-package-alt; is there a way we could reuse instead of copy?